### ISSUE TITLE
gpg: decryption failed: No secret key

### ISSUE THREAD

**Issue by aymericbeaumet** [StackExchange/blackbox#234]
Blackbox: [89566f7dd298449356f295ce5e841aae9bd028d0](https://github.com/StackExchange/blackbox/tree/89566f7dd298449356f295ce5e841aae9bd028d0)
Gpg: 2.2.3

We are facing a Blackbox issue on both several local machines and the CI (CircleCI). We extensively rely on Blackbox with nearly a 100 encrypted files.

The decryption goes well 95% of the time, but the rest of a time we are facing a non-deterministic issue causing gpg to crash, here goes the relevant part of the error logs:

```
$ blackbox_decrypt_all_files
gpg: key ABCD: public key "abcd <abcd@key>" imported
gpg: key ABCD: secret key imported
gpg: Total number processed: 1
gpg:               imported: 1
gpg:       secret keys read: 1
gpg:   secret keys imported: 1
========== Importing keychain: START
gpg: key 98A87F81DCF732EA: public key "a <a@a>" imported
gpg: key DFB3BB7E72A6AE86: public key "b <b@b>" imported
gpg: key E718D80E5FACEB2B: public key "c <c@c>" imported
gpg: key EFAF45AA074BFF85: public key "d <d@d>" imported
gpg: key E6D589AD1CBB467F: public key "e <e@e>" imported
gpg: key CB36B534CB358125: public key "f <f@f>" imported
gpg: key F591723101148D38: public key "g <g@g>" imported
gpg: Total number processed: 8
gpg:               imported: 7
gpg:              unchanged: 1
========== Importing keychain: DONE
========== Decrypting new/changed files: START
========== EXTRACTED some/secret/file.js
# ...
# Successfully extracting ~90 files
# ...
========== EXTRACTED some/other/secret/file.js
gpg: decryption failed: No secret key
Exited with code 2
```

Hard to track down, any ideas?

**Comment by TomOnTime** [comment]
If you get "gpg: decryption failed: No secret key" it often means a file was not re-encrypted  with the new key. Does that help?

**Comment by aymericbeaumet** [comment]
I'm pretty sure the the files are properly encrypted as Blackbox might fail the first time it is being invoked, but succeed the second time (without any change to the files). I'm trying to guess why that non-deterministic behavior is happening.

You never had a similar issue at SE?

**Comment by tlimoncelli** [comment]
Sadly (?) we've never seen this behavior at SE.  We do a blackbox_decrypt_all_files in our CI system many times each day.  I'd love to track this down since it seems like an edge-case that could bite others.

**Comment by aymericbeaumet** [comment]
I've been trying to find a repro (on Linux/macOS), but did not succeed yet. Was hoping to find someone having the same problem when opening this issue.

Maybe it is related to the number of files we encrypt with Blackbox? I'm wondering whether it's putting too much pressure on the gpg daemon.

**Comment by tlimoncelli** [comment]
I hadn't considered that. If it helps any... at SE, our repo with the most BB files has 61. The others have 2-3 files.

You could test your hypothesis pretty easily: put a "sleep 1" in the loop.  That will slow it down considerably (not that you'd want it always that way).

**Comment by aymericbeaumet** [comment]
I cannot reproduce the bug locally on OSX, but I found a way on CircleCI (with the aforementioned base image).

## First attempt

_Hypothesis_: corrupted gpg-agent / missing env var

_Attempt_: kill the gpg-agent / declare the `GPG_TTY` env var

```
for i in {1..9999} ; do
  echo "TRY: $i"; git clean -ffdx
  export GPG_TTY=$(tty); gpgconf --kill gpg-agent; killall gpg-agent
  blackbox_decrypt_all_files || break
done
```

## Second attempt

_Hypothesis_: too much requests are being sent to the gpg-agent

_Attempt_: delay each loop turn in [blackbox_postdeploy](https://github.com/StackExchange/blackbox/blob/master/bin/blackbox_postdeploy#L30-L38) (/cc @TomOnTime)

_The loop is the same as for the first attempt._

- 0.25s 🔴
- 0.5s 🔴

It takes longer but the decrypt eventually always fails.

I could not manage to make it fail with a 1s sleep (after ~30 minutes). Maybe I have not been patient enough.

## Temporary fix

This is not perfect, but retrying a few times at least makes the build work.

```bash
for i in {1..3}; do
  blackbox_decrypt_all_files && exit 0
done
exit 1
```

## Next steps

Investigate and try `decrypt_file_overwrite`.

## Related readings

- https://stackoverflow.com/a/42472770/1071486
- https://superuser.com/questions/1075404/how-can-i-restart-gpg-agent
- https://stackoverflow.com/questions/28321712/gpg-decryption-fails-with-no-secret-key-error/42472770#42472770

**Comment by jbonzo** [comment]
I've noticed similar problems recently, is this issue still relevant for @aymericbeaumet ?

**Comment by aymericbeaumet** [comment]
Yeah we are still facing this issue, and have been staying with the _temporary_ ever since

**Comment by rrggrr** [comment]
Issue persists

**Comment by coding-bunny** [comment]
I'm running into a very similar issue, especially when swapping between branches and keys getting added/removed as well from new employees.
To give a more detailed overview of the problem:

* I'm able to successfully decrypt all files on the master branch using `blackbox_decrypt_all_files`
* This command doesn't work on my branch, failing with the `gpg: decryption failed: No secret key`
* Copied over the files from master branch
* Can decrypt successfully
* made changes and re-encrypted the files
* pushed to git
* `blackbox_decrypt_all_files` stops working again with the same error
* still works fine on the master branch

And this happens to various projects, and we're talking about maybe 5 files maximum  being encrypted in these projects.
Killing the GPG agent sometimes helps but now always.

**Comment by tlimoncelli** [comment]
Interesting new issue!

Is this a problem with GPG or with Blackbox?  Do you think gpg-agent is getting confused by git replacing so many files?

**Comment by coding-bunny** [comment]
I honestly don't know.
I can for example fire up `gpg --edit-key <KEYID>` and this works fine.
But running decrypt commands or blackbox commands tells me that the wrong key is being used somehow.

I'm trying to figure out if gpg is not using something from the macOS keychain or something?
I'm really running out of ideas at the moment.

The weird part is that I can work just fine other branches, or by copying the files.
But as soon as I make a change everything breaks.

**Comment by coding-bunny** [comment]
resolved the issue.
The problem is caused by old public keys, that no longer have the private key available.
Once we cleaned up those keys it all worked again

**Comment by isaiahinocentes** [comment]
Had the same problem. I installed blackbox using debian and tried to decrypt the files. I used curses for my pinentry and restarted/killed the gpg agent then it worked for me. 

My guess was I needed to restart my gpg agent when I changed my pinenty in the gpg-agent.conf

gpg in command line using pinentry-curses:
https://chaosfreakblog.wordpress.com/2013/06/21/gpg-problem-with-the-agent-no-pinentry-solved/

and also thanks to this thread as for references.

Thanks

**Comment by coding-bunny** [comment]
I actually ran into a curious problem again today.
The blackbox failed to decrypt again, and even the normal `gpg -d` command didn't work after exchanging the keys.

I removed all duplicate entries on my system for invalid keys and noticed that multiple `gpg-agents` had been spawned. Once I killed those everything started to work again properly.
It seems blackbox has some issue when multiple agents are spawned and keys are replaced for an admin. As if the wrong keys are being used for encrypting, or not all public known keys are used or something.

**Comment by mig8447** [comment]
Just in case someone is going through the same error and none of the solutions above worked:

I had an Ubuntu 18.04 machine with the Gnome GUI and by default, `gpg` was using `/usr/bin/pinentry-gnome3`, which is a GUI-only pinentry not able to be accessed through SSH, which is exactly what i was doing. To fix it:

1. Install `pinentry-tty` by executing (In Ubuntu's case):
    ```bash
    sudo apt-get install pinentry-tty
    ```
2. Set the GPG Agent's pinentry program to `pinentry-tty` by editing (or creating) the `~/.gnupg/gpg-agent.conf` file and adding the following line:
    ```
    pinentry-program /usr/bin/pinentry-tty
    ```
3. Reload the GPG Agent by executing
    ```bash
    gpg-connect-agent reloadagent /bye
    ```

References:
- https://superuser.com/a/521027/968840
